### DISCUSSION TITLE
AppDelegate returning handled state in methods

### DISCUSSION THREAD

**Discussion by chanki31** [pointfreeco/swift-composable-architecture#519] [Q&A]
Consider the following
```
extension AppDelegate {
    func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey: Any] = [:]) -> Bool {
        viewStore.send(.appDelegate(.open(url)))
        return true
    }
}
```

I'm trying to find a way to return the handled state when the above method is called by the system using ComposableArchitecture. Any thoughts?

**Comment by mbrandonw** [comment]
One possibility is to hold the state that you want to return in your domain (I think in this case it would be `AppState`), mutate that state when `.open(url)` action is sent, and then access that state after sending the action. This is possible since sending actions is synchronous, which we outline a few reasons why this is in our [FAQ](https://github.com/pointfreeco/swift-composable-architecture#faq) but this is another good reason to do it.

Here's some sample code:

```swift
struct AppState: Equatable {
  ...
  var didOpenUrl = false
}

enum AppAction {
  ...
  case open(URL)
}

let appReducer = Reducer<AppState, AppAction, AppEnvironment> { state, action, environment in 
  switch action {
  ...

  case let .open(url):
    // Logic to handle url
    // ...
    state.didOpenUrl = true // or false if url was not handled
    return .none
  }
}

extension AppDelegate {
  func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey: Any] = [:]) -> Bool {
    viewStore.send(.open(url))
    return viewStore.didOpenUrl
  }
}
```

**Reply by chanki31** [reply]
I must have missed the point where sending actions are sync, thanks for this much appreciated
